rust/kernel/rbtree.rs

```diff
@@ -418,6 +602,221 @@ impl<K, V> Drop for RBTree<K, V> {
+    /// tree.try_insert(20, 200)?;
+    /// tree.try_insert(30, 300)?;
+    ///
+    /// // remove the first element
+    /// let mut cursor = tree.cursor_front().unwrap();
+    /// let mut current = cursor.current();
+    /// assert_eq!(current, (&10, &100));
+    /// cursor = cursor.remove_current().unwrap();
+    /// current = cursor.current();
+    /// assert_eq!(current, (&20, &200));
+    ///
+    /// # Ok::<(), Error>(())
+    /// ```
+    /// Returns a cursor to the previous node if there is no next node
+    /// ```
+    /// use kernel::rbtree::RBTree;
+    ///
+    /// // Create a new tree.
+    /// let mut tree = RBTree::new();
+    ///
+    /// // Insert three elements.
+    /// tree.try_insert(10, 100)?;
+    /// tree.try_insert(20, 200)?;
+    /// tree.try_insert(30, 300)?;
+    ///
+    /// // remove the last element
+    /// let mut cursor = tree.cursor_back().unwrap();
+    /// let mut current = cursor.current();
+    /// assert_eq!(current, (&30, &300));
+    /// cursor = cursor.remove_current().unwrap();
+    /// current = cursor.current();
+    /// assert_eq!(current, (&20, &200));
+    ///
+    /// # Ok::<(), Error>(())
+    /// ```
+    /// Returns None if the current node was the last one in the tree
+    /// ```
+    /// use kernel::rbtree::RBTree;
+    ///
+    /// // Create a new tree.
+    /// let mut tree = RBTree::new();
+    ///
+    /// // Insert one elements.
+    /// tree.try_insert(20, 200)?;
+    ///
+    /// // remove the only element
+    /// let mut cursor = tree.cursor_front().unwrap();
+    /// let current = cursor.current();
+    /// assert_eq!(current, (&20, &200));
+    /// let cursor = cursor.remove_current();
+    /// assert!(cursor.is_none());
+    ///
+    /// # Ok::<(), Error>(())
+    /// ```
+    pub fn remove_current(&mut self) -> Option<Self> {
+        let prev = self.get_neighbor_raw(Direction::Prev);
+        let next = self.get_neighbor_raw(Direction::Next);
+        let this = crate::container_of!(self.current, Node<K, V>, links) as *mut Node<K, V>;
+        // SAFETY: The reference to the tree used to create the cursor outlives the cursor, so
+        // the tree cannot change. By the tree invariant, all nodes are valid.
+        unsafe { bindings::rb_erase(&mut (*this).links, self.root) };
+
+        let current = match (prev, next) {
+            (_, Some(next)) => next,
+            (Some(prev), None) => prev,
+            (None, None) => {
+                return None;
+            }
+        };
+
+        Some(Self {
+            current,
+            _tree: self._tree,
+            root: self.root,
+        })
+    }
+
+    /// Move the cursor to the previous node,
+    /// returning None if it doesn't exist
+    pub fn prev(&mut self) -> Option<Self> {
+        self.mv(Direction::Prev)
+    }
+
+    /// Move the cursor to the next node,
+    /// returning None if it doesn't exist
+    pub fn next(&mut self) -> Option<Self> {
+        self.mv(Direction::Next)
+    }
```

Comment:
These must destroy the cursor they are called on.
```suggestion
    /// Move the cursor to the previous node,
    /// returning None if it doesn't exist
    pub fn prev(self) -> Option<Self> {
        self.mv(Direction::Prev)
    }

    /// Move the cursor to the next node,
    /// returning None if it doesn't exist
    pub fn next(self) -> Option<Self> {
        self.mv(Direction::Next)
    }
```

Reply:
Is this accomplished by just taking an owned self instead of &self? Or does it have to be explicitly dropped

Reply:
Nevermind, didn't realize the suggestion was already there :)

Reply:
used `mut self` here and on `remove_current`.  PLMK if that is wrong